# Problem 3

Given a 3rd degree polynomial:

$
f(x) = ax^3 + bx^2 + cx + d
\renewcommand{\R}{\mathbb{R}}
$
 such that $a,b,c,d \in \R$

Because the function is cubic, f(x) satisfies Property 1 of cubic splines.

Since f(x) is also in cubic format, f(x) is able to interpolate through all data nodes. I.e. $ \forall x_i \in [x_0,x_n], f(x) = s(x)$ such that $s(x)$ is a cubic spline of $f(x)$.

f(x) satisfies Property 2 of cubic splines.

Because f(x) is a polynomial, f, f', and f'' are all continuous.

This satisfies Property 3, 4, and 5 of cubic splines.

For Property 6, we know that $S'(x_0) = f'(x_0)$ has to be true. 

$
f'(x) = 3ax^2 + 2bx + c
$

So, let S'(x) = f'(x), and this implies that f(x) can satisfy its own clamped spline condition.

However, checking the natural spline condition, if $S(x) = f(x)$, then for $f(x)$ to be natural, $S''(x_0) = S''(x_n) = 0$.

$
S''(x_0) = 6ax_0 + 2b \\
6ax_0 + 2b \implies x_0 = -\frac{b}{3a} \\
6ax_n + 2b \implies x_n = -\frac{b}{3a} 
$

This implies that $x_0$ and $x_n$ are not a 0 of S''(x), which implies that f(x) cannot satisfy the natural spline condition.

# Problem 4

## 3a

In [65]:
import numpy as np

# only 2 nodes, so h is 1 value
h = 0.3

# given output is ai
ai = np.array([17.56492,18.50515])

# given x_0 -> x_n
xi = np.array([8.3,8.6])

# since 2 nodes, A is identity
A = np.mat([1,0,0,1]).reshape(2,2)

# since 2 nodes, z array is 0
z = np.zeros(2)

# find ci, n x 1 dim
ci = np.linalg.solve(A, z)

# bi using a and h and ci
bi = np.array([])
for i in range(len(ai)-1):
    res = ((ai[i+1] - ai[i])/(h)) - ((h * (ci[i+1] + 2 * ci[i])/3))
    bi = np.append(bi,res)

# di using ci
di = np.array([])
for i in range(len(ci)-1):
    res = (ci[i+1]-ci[i])/(3*h)
    di = np.append(di,res)

# print result
print(ai[0],bi,ci,di)

17.56492 [3.1341] [0. 0.] [0.]


The results imply that the cubic spline in the interval $[x_0,x_1]$ is the following equation:

$
S_0(x) = 17.56492 + 3.1341x
$
 such that $8.3 \leq x < 8.6$

## 5a

In [66]:
func = lambda x : x * np.log(x)
sfunc = lambda x : 17.56492 + 3.1341*x
print(sfunc(8.4) - func(8.4))

26.014213670866155


In [67]:
func_deriv = lambda x : np.log(x) + 1
sfunc_deriv = 3.1341
print(func_deriv(8.4) - sfunc_deriv)

-0.005868294150732201


The results imply that the actual error between $f(x)$ and $S(x)$ is about 26.0142. The actual error between $f'(x)$ and $S'(x)$ is about -0.00587.